cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/C.hs

```diff
@@ -0,0 +1,618 @@
+  c_blst_from_affine = c_blst_p2_from_affine
+  c_blst_affine_in_g = c_blst_p2_affine_in_g2
+
+  c_blst_generator = c_blst_p2_generator
+
+  c_blst_x_from_p = c_blst_x_from_p2
+  c_blst_y_from_p = c_blst_y_from_p2
+  c_blst_z_from_p = c_blst_z_from_p2
+
+  c_blst_x_from_affine = c_blst_x_from_affine2
+  c_blst_y_from_affine = c_blst_y_from_affine2
+
+  c_blst_p_is_equal = c_blst_p2_is_equal
+  c_blst_p_is_inf = c_blst_p2_is_inf
+
+---- Safe Scalar types / marshalling
+
+sizeScalar :: forall i. Num i => i
+sizeScalar = fromIntegral c_size_blst_scalar
+
+newtype Scalar = Scalar (ForeignPtr Void)
+
+withScalar :: Scalar -> (ScalarPtr -> IO a) -> IO a
+withScalar (Scalar p2) go = do
+  withForeignPtr p2 (go . ScalarPtr)
+
+withNewScalar :: (ScalarPtr -> IO a) -> IO (a, Scalar)
+withNewScalar go = do
+  p2 <- mallocForeignPtrBytes sizeScalar
+  x <- withForeignPtr p2 (go . ScalarPtr)
+  return (x, Scalar p2)
+
+withNewScalar_ :: (ScalarPtr -> IO a) -> IO a
+withNewScalar_ = fmap fst . withNewScalar
+
+withNewScalar' :: (ScalarPtr -> IO a) -> IO Scalar
+withNewScalar' = fmap snd . withNewScalar
+
+cloneScalar :: Scalar -> IO Scalar
+cloneScalar (Scalar a) = do
+  b <- mallocForeignPtrBytes sizeScalar
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap sizeScalar
+  return (Scalar b)
+
+sizeFr :: forall i. Num i => i
+sizeFr = fromIntegral c_size_blst_fr
+
+newtype Fr = Fr (ForeignPtr Void)
+
+withFr :: Fr -> (FrPtr -> IO a) -> IO a
+withFr (Fr p2) go = do
+  withForeignPtr p2 (go . FrPtr)
+
+withNewFr :: (FrPtr -> IO a) -> IO (a, Fr)
+withNewFr go = do
+  p2 <- mallocForeignPtrBytes sizeFr
+  x <- withForeignPtr p2 (go . FrPtr)
+  return (x, Fr p2)
+
+withNewFr_ :: (FrPtr -> IO a) -> IO a
+withNewFr_ = fmap fst . withNewFr
+
+withNewFr' :: (FrPtr -> IO a) -> IO Fr
+withNewFr' = fmap snd . withNewFr
+
+cloneFr :: Fr -> IO Fr
+cloneFr (Fr a) = do
+  b <- mallocForeignPtrBytes sizeFr
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap sizeFr
+  return (Fr b)
+
+---- Safe FP types / marshalling
+
+sizeFP :: forall i. Num i => i
+sizeFP = fromIntegral c_size_blst_scalar
+
+newtype FP = FP (ForeignPtr Void)
+
+withFP :: FP -> (FPPtr -> IO a) -> IO a
+withFP (FP p2) go = do
+  withForeignPtr p2 (go . FPPtr)
+
+withNewFP :: (FPPtr -> IO a) -> IO (a, FP)
+withNewFP go = do
+  p2 <- mallocForeignPtrBytes sizeFP
+  x <- withForeignPtr p2 (go . FPPtr)
+  return (x, FP p2)
+
+withNewFP_ :: (FPPtr -> IO a) -> IO a
+withNewFP_ = fmap fst . withNewFP
+
+withNewFP' :: (FPPtr -> IO a) -> IO FP
+withNewFP' = fmap snd . withNewFP
+
+cloneFP :: FP -> IO FP
+cloneFP (FP a) = do
+  b <- mallocForeignPtrBytes sizeFP
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap sizeFP
+  return (FP b)
+
+fpToNat :: FP -> IO Natural
+fpToNat fp = withFP fp $ \fpPtr -> do
+  allocaBytes 48 $ \rawPtr -> do
+    c_blst_bendian_from_fp rawPtr fpPtr
+    cstrToNat rawPtr 48
```

Comment:
The hardcoded lengths of 48 could be replaced with `sizeFP` for consistency

Reply:
Module has been removed, we will not be needing this.